### PR TITLE
paraflow: automatic email invitation to para

### PR DESCRIPTION
**Pull request by BrettEastman**
- Add Nodemailer module for sending email, located at src/config/nodemailer.ts
- Add logic to src/backend/routers/paras.ts to send email at the same time para email first entered into db by cm

### CODE REVIEW

File: src/backend/routers/paras.ts [hunk 1]
```diff
@@ -34,13 +35,23 @@ export const paraProcedures = {
     )
     .mutation(async (req) => {
       const { first_name, last_name, email, role } = req.input;
+      // In the SCHEMA, email has the "unique" constraint TEXT UNIQUE NOT NULL, which will prevent duplicate paras from being inputted
+
+      // this is where Nodemailer will send an email to the email address of the para inputted by the cm
+      await transporter.sendMail({
```
Comment (codetheweb) [inline src/backend/routers/paras.ts:41]: might make more sense to send the email after inserting to our database, don't want to send an email if the insert fails for some reason

File: src/components/case_manager/AllParas.tsx [hunk 2]
```diff
@@ -21,6 +21,7 @@ const AllParasPage = () => {
       email: data.get("email") as string,
       role: "para",
     });
+    // email invitation here? Or should be in the backend as part of mutate?
```
Comment (codetheweb) [inline src/components/case_manager/AllParas.tsx:24]:
```suggestion
```

File: src/config/nodemailer.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { createTransport } from "nodemailer";
+
+const email: string | undefined = process.env.EMAIL;
+const pass: string | undefined = process.env.EMAIL_PASS;
```
Comment (codetheweb) [inline src/config/nodemailer.ts:4]: these should be added to `src/backend/lib/types/env.ts`

also would be good to document them in `.env.example`

File: src/components/case_manager/AllParas.tsx (hunk 2 above)
Comment (codetheweb) [inline src/components/case_manager/AllParas.tsx:24]:
```suggestion
```

**Comment by codetheweb** [comment]
also would be good to add a test, could be as simple as adding an event handler on the node mailer object and updating a global state variable with sent messages when `process.env.NODE_ENV === "test"`

File: src/backend/routers/paras.ts [hunk 4]
```diff
@@ -34,13 +35,23 @@ export const paraProcedures = {
...
+        ...mailOptions,
+        to: email,
+        subject: "Para-professional email confirmation",
+        text: "Email confirmation",
+        html: "<h1>Email confirmation</h1><p>Please confirm your email by going to the following link: <a>no link yet</a></p>",
```
Comment (amantri) [inline src/backend/routers/paras.ts:46]: Please add a TODO to insert link later

File: src/config/nodemailer.ts
Comment (amantri) [inline src/config/nodemailer.ts:1]: Not sure this is config - maybe lib/nodemailer.ts would be a better home for this file.

File: src/backend/routers/paras.ts (hunk 1 above)
Comment (BrettEastman) [reply src/backend/routers/paras.ts:47]: Hi Max, in paras.ts, I moved transporter.sendMail to after .execute. I'm not sure if that is what you meant. Please let me know.

Comment (codetheweb) [reply src/backend/routers/paras.ts:47]: looks good!

File: src/backend/lib/nodemailer.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    user: email,
+    pass,
+  },
+});
+
+// this just means we are sending an email from our own email for testing purposes. The "to" email will be changed in src/backend/routers/paras to the para email address entered into the db by the cm.
+export const mailOptions = {
+  from: email,
+  to: email,
+};
```
Comment (codetheweb) [inline src/backend/lib/nodemailer.ts:18]: is this needed since the only parameter that's used here is `from`?

File: src/backend/routers/paras.ts [hunk 6]
```diff
@@ -34,13 +35,25 @@ export const paraProcedures = {
...
+      // In the SCHEMA, email has the "unique" constraint TEXT UNIQUE NOT NULL, which will prevent duplicate paras from being inputted
 
-      // todo: add a unique constraint to prevent duplicate paras
       const result = await req.ctx.db
         .insertInto("user")
         .values({ first_name, last_name, email, role })
         .returningAll()
         .execute();
+
+      // this is where Nodemailer will send an email to the email address of the para inputted by the cm
```
Comment (codetheweb) [inline src/backend/routers/paras.ts:46]:
```suggestion
```

File: src/backend/routers/paras.ts [hunk 7]
```diff
@@ -34,13 +35,25 @@ export const paraProcedures = {
...
+      await transporter.sendMail({
+        ...mailOptions,
+        to: email,
+        subject: "Para-professional email confirmation",
+        text: "Email confirmation",
+        html: "<h1>Email confirmation</h1><p>Please confirm your email by going to the following link: <a>no link yet</a></p>",
+      });
+      // to do here: when site is deployed, add url to html above
+      // to do elsewhere: add "email_verified_at" timestamp when para first signs in with their email address (entered into db by cm)
```
Comment (codetheweb) [inline src/backend/routers/paras.ts:55]: could be wrong but I think this is already done by our auth adapter

File: src/backend/routers/paras.ts [hunk 8]
```diff
@@ -34,13 +35,25 @@ export const paraProcedures = {
     )
     .mutation(async (req) => {
       const { first_name, last_name, email, role } = req.input;
+      // In the SCHEMA, email has the "unique" constraint TEXT UNIQUE NOT NULL, which will prevent duplicate paras from being inputted
```
Comment (codetheweb) [inline src/backend/routers/paras.ts:38]:
```suggestion
```

File: src/backend/lib/nodemailer.ts (hunk 5 above)
Comment (BrettEastman) [reply src/backend/lib/nodemailer.ts:18]: Fixed

**Review by codetheweb — APPROVED** [review]
would be nice to get a test in but otherwise looks good!

**Pull request merged**